db/sqldb/incomplete_upload.go

```diff
@@ -0,0 +1,223 @@
+		return nil, fmt.Errorf("translating rows to incomplete_uploads: %w", err)
+	}
+	return exactlyOne("incomplete_upload", id, pvs)
+}
+
+func (d *DB) IncompleteUploads(tx db.Tx, ids []pacta.IncompleteUploadID) (map[pacta.IncompleteUploadID]*pacta.IncompleteUpload, error) {
+	ids = dedupeIDs(ids)
+	rows, err := d.query(tx, `
+		SELECT `+incompleteUploadSelectColumns+`
+		FROM incomplete_upload 
+		WHERE id IN `+createWhereInFmt(len(ids))+`;`, idsToInterface(ids)...)
+	if err != nil {
+		return nil, fmt.Errorf("querying incomplete_uploads: %w", err)
+	}
+	pvs, err := rowsToIncompleteUploads(rows)
+	if err != nil {
+		return nil, fmt.Errorf("translating rows to incomplete_uploads: %w", err)
+	}
+	result := make(map[pacta.IncompleteUploadID]*pacta.IncompleteUpload, len(pvs))
+	for _, pv := range pvs {
+		result[pv.ID] = pv
+	}
+	return result, nil
+}
+
+func (d *DB) CreateIncompleteUpload(tx db.Tx, i *pacta.IncompleteUpload) (pacta.IncompleteUploadID, error) {
+	if err := validateIncompleteUploadForCreation(i); err != nil {
+		return "", fmt.Errorf("validating incomplete_upload for creation: %w", err)
+	}
+	hd, err := encodeHoldingsDate(i.HoldingsDate)
+	if err != nil {
+		return "", fmt.Errorf("validating holdings date: %w", err)
+	}
+	i.ID = pacta.IncompleteUploadID(d.randomID("iu"))
+	err = d.exec(tx, `
+		INSERT INTO incomplete_upload 
+			(id, owner_id, admin_debug_enabled, blob_id, name, description, holdings_date)
+			VALUES
+			($1, $2, $3, $4, $5, $6, $7);`,
+		i.ID, i.Owner.ID, i.AdminDebugEnabled, i.Blob.ID, i.Name, i.Description, hd)
+	if err != nil {
+		return "", fmt.Errorf("creating incomplete_upload: %w", err)
+	}
+	return i.ID, nil
+}
+
+func (d *DB) UpdateIncompleteUpload(tx db.Tx, id pacta.IncompleteUploadID, mutations ...db.UpdateIncompleteUploadFn) error {
+	err := d.RunOrContinueTransaction(tx, func(db.Tx) error {
+		init, err := d.IncompleteUpload(tx, id)
+		if err != nil {
+			return fmt.Errorf("reading incomplete_upload: %w", err)
+		}
+		for i, m := range mutations {
+			err := m(init)
+			if err != nil {
+				return fmt.Errorf("running %d-th mutation: %w", i, err)
+			}
+		}
+		err = d.putIncompleteUpload(tx, init)
+		if err != nil {
+			return fmt.Errorf("putting incomplete_upload: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("updating incomplete_upload: %w", err)
+	}
+	return nil
+}
+
+func (d *DB) DeleteIncompleteUpload(tx db.Tx, id pacta.IncompleteUploadID) (pacta.BlobURI, error) {
+	var buri pacta.BlobURI
+	err := d.RunOrContinueTransaction(tx, func(tx db.Tx) error {
+		iu, err := d.IncompleteUpload(tx, id)
+		if err != nil {
+			return fmt.Errorf("reading incomplete_upload: %w", err)
+		}
+		err = d.exec(tx, `DELETE FROM incomplete_upload WHERE id = $1;`, id)
+		if err != nil {
+			return fmt.Errorf("deleting incomplete_upload: %w", err)
+		}
+		buri, err = d.DeleteBlob(tx, iu.Blob.ID)
+		if err != nil {
+			return fmt.Errorf("deleting blob: %w", err)
+		}
+		return nil
+	})
```

Comment:
This could also be a single `DELETE FROM incomplete_upload WHERE id = $1 RETURNING blob_id` instead of three separate database calls, though I care less about this one than the one deleting N child rows.

Reply:
Yeah good thought - here I think the argument cuts the other way (see above) so I'll skip.